File: DirectShowFilters/TsWriter/source/LvctParser.cpp
/*
* Copyright (C) 2006-2008 Team MediaPortal
* http://www.team-mediaportal.com
*
* This Program is free software; you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation; either version 2, or (at your option)
* any later version.
*
* This Program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with GNU Make; see the file COPYING. If not, write to
* the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
* http://www.gnu.org/copyleft/gpl.html
*
*/
#pragma warning(disable : 4995)
#include <windows.h>
#include <bdatypes.h>
#include "..\..\shared\BasePmtParser.h"
#include "LvctParser.h"

extern void LogDebug(const char *fmt, ...);
extern bool DisableCRCCheck();

CLvctParser::CLvctParser(int pid)
{
  SetPid(pid);
  if (DisableCRCCheck())
  {
    EnableCrcCheck(false);
  }
  Reset();
  m_pCallBack = NULL;
}

CLvctParser::~CLvctParser(void)
{
}

void CLvctParser::SetCallBack(ILvctCallBack* callBack)
{
  m_pCallBack = callBack;
}

bool CLvctParser::IsReady()
{
  return m_bIsReady;
}

void CLvctParser::Reset()
{
  LogDebug("LvctParser: reset");
  CSectionDecoder::Reset();
  m_iCurrentVersion = -1;
  m_mUnseenSections.clear();
  m_bIsReady = false;
  LogDebug("LvctParser: reset done");
}

void CLvctParser::OnNewSection(CSection& sections)
{
  // 0xc8 = terrestrial virtual channel table
  // 0xc9 = cable virtual channel table
  if ((sections.table_id != 0xc8 && sections.table_id != 0xc9) || m_bIsReady || m_pCallBack == NULL)
  {
    return;
  }
  byte* section = sections.Data;

  try
  {
    int sectionSyntaxIndicator = section[1] & 0x80;
    int sectionLength = ((section[1] & 0xf) << 8) + section[2];
    if ((sections.table_id == 0xc8 && sectionLength > 1021) || sectionLength > 4093 || sectionLength < 13)
    {
      LogDebug("LvctParser: invalid section length = %d", sectionLength);
      return;
    }
    int transportStreamId = (section[3] << 8) + section[4];
    int versionNumber = (section[5] >> 1) & 0x1f;
    int currentNextIndicator = section[5] & 1;
    if (currentNextIndicator == 0)
    {
      // Details do not apply yet...
      return;
    }
    int sectionNumber = section[6];
    int lastSectionNumber = section[7];

    int endOfSection = sectionLength - 1;
    LogDebug("LvctParser: table ID = 0x%x, transport stream ID = 0x%x, version = %d, section number = %d, last section number = %d, section length = %d, end of section = %d",
      sections.table_id, transportStreamId, versionNumber, sectionNumber, lastSectionNumber, sectionLength, endOfSection);

    if (versionNumber > m_iCurrentVersion || (versionNumber == MAX_TABLE_VERSION_NUMBER && versionNumber < m_iCurrentVersion))
    {
      LogDebug("LvctParser: new table version, table ID = 0x%x, version = %d, section number = %d, last section number = %d", sections.table_id, versionNumber, sectionNumber, lastSectionNumber);
      m_iCurrentVersion = versionNumber;
      m_mUnseenSections.clear();
      for (int s = 0; s <= lastSectionNumber; s++)
      {
        m_mUnseenSections.push_back(s);
      }
    }
    vector<int>::iterator sectionIt = find(m_mUnseenSections.begin(), m_mUnseenSections.end(), sectionNumber);
    if (sectionIt == m_mUnseenSections.end())
    {
      return;
    }

    int protocolVersion = section[8];
    int numChannelsInSection = section[9];
    //LogDebug("LvctParser: protocol version = %d, number of channels = %d", protocolVersion, numChannelsInSection);
    int pointer = 10;
    for (int i = 0; i < numChannelsInSection && pointer + 31 < endOfSection; i++)
    {
      // shortName = 7*16 bits (14 bytes), UTF-16 encoding
      char* name = new char[8];
      if (name == NULL)
      {
        LogDebug("LvctParser: failed to allocate 8 bytes for the short name");
      }
      else
      {
        for (int count = 0; count < 7; count++)
        {
          // Take every second byte as a rough ASCII conversion.
          pointer++; // skip the UTF-16 high byte
          name[count] = section[pointer++];
          name[count + 1] = 0; // NULL terminate
        }
        LogDebug("LvctParser: short name = %s", name);
      }

      // SCTE cable supports both one-part and two-part channel numbers where
      // the major and minor channel number range is 0..999.
      // ATSC supports only two-part channel numbers where the major range is
      // 1..99 and the minor range is 0..999.
      // When the minor channel number is 0 it indicates an analog channel.
      int majorChannel = ((section[pointer] & 0xf) << 6) + (section[pointer + 1] >> 2);
      pointer++;
      int minorChannel = ((section[pointer] & 0x3) << 8) + section[pointer + 1];
      pointer += 2;
      if (((majorChannel >> 4) & 0x3f) == 0x3f)
      {
        majorChannel = ((majorChannel & 0xf) << 10) + minorChannel;
        minorChannel = 0;
      }

      // Frequency and modulation might be useful as part of an ATSC network
      // scan... except that the standard says that the frequency is deprecated
      // and should be set to zero. There is no other known ATSC NIT standard.
      int modulationMode = section[pointer++];
      unsigned int carrierFrequency = section[pointer++] << 24;  // Hz
      carrierFrequency += (section[pointer++] << 16);
      carrierFrequency += (section[pointer++] << 8);
      carrierFrequency += section[pointer++];

      int channelTsid = ((section[pointer]) << 8) + section[pointer + 1];
      pointer += 2;
      int programNumber = ((section[pointer]) << 8) + section[pointer + 1];
      pointer += 2;
      int etmLocation = (section[pointer] >> 6);
      bool accessControlled = ((section[pointer] >> 5) & 0x1);
      bool hidden = ((section[pointer] >> 4) & 0x1);

      // cable only
      int pathSelect = 0;
      bool outOfBand = false;
      if (sections.table_id == 0xc9)
      {
        pathSelect = ((section[pointer] >> 3) & 0x1);
        outOfBand = ((section[pointer] >> 2) & 0x1);
      }

      bool hideGuide = ((section[pointer++] >> 1) & 0x1);
      int serviceType = (section[pointer++] & 0x3f);
      int sourceId = ((section[pointer]) << 8) + section[pointer + 1];
      pointer += 2;
      LogDebug("LvctParser: major channel = %d, minor channel = %d, modulation = %d, carrier frequency = %d, channel TSID = 0x%x, program number = 0x%x, ETM location = %d, access controlled = %d, hidden = %d, path select = %d, out of band = %d, hide guide = %d, service type = 0x%x, source ID = 0x%x",
                  majorChannel, minorChannel, modulationMode, carrierFrequency, channelTsid, programNumber,
                  etmLocation, accessControlled, hidden, pathSelect, outOfBand, hideGuide, serviceType, sourceId);

      // descriptors...
      int descriptorsLength = ((section[pointer] & 0x3) << 8) + section[pointer + 1];
      pointer += 2;
      int endOfDescriptors = pointer + descriptorsLength;
      LogDebug("LvctParser: pointer = %d, descriptors length = %d, end of descriptors = %d", pointer, descriptorsLength, endOfDescriptors);

      int videoStreamCount = 0;
      int audioStreamCount = 0;
      vector<unsigned int> languages;
      while (pointer + 1 < endOfDescriptors)
      {
        int tag = section[pointer++];
        int length = section[pointer++];
        LogDebug("LvctParser: descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d", tag, length, pointer, pointer + length);
        if (pointer + length > endOfDescriptors)
        {
          LogDebug("LvctParser: invalid descriptor length = %d, pointer = %d, end of descriptor = %d, end of descriptors = %d, end of section = %d, section length = %d", length, pointer, pointer + length, endOfDescriptors, endOfSection, sectionLength);
          if (name != NULL)
          {
            delete[] name;
            name = NULL;
          }
          return;
        }

        if (tag == 0xa0) // extended channel name descriptor
        {
          vector<char*> extendedNames;
          DecodeMultipleStrings(&section[pointer], length, &extendedNames);
          // Combine the first different extended name with the short name... carefully!
          int shortNameLength = 0;
          if (name != NULL)
          {
            shortNameLength = strlen(name);
          }
          for (size_t i = 0; i < extendedNames.size(); i++)
          {
            int extendedNameLength = strlen(extendedNames[i]);
            if (extendedNameLength == shortNameLength && strcmp(name, extendedNames[0]) == 0)
            {
              continue;
            }
            int nameBufferSize = 0;
            if (name != NULL)
            {
              nameBufferSize += strlen(" ()") + shortNameLength;
            }
            nameBufferSize += extendedNameLength + 1; // + 1 for NULL termination
            char* newName = new char[nameBufferSize];
            if (newName == NULL)
            {
              LogDebug("LvctParser: failed to allocate %d bytes for the extended name", nameBufferSize);
            }
            else
            {
              strcpy(newName, extendedNames[i]);
              if (name != NULL)
              {
                strcat(newName, " (");
                strcat(newName, name);
                strcat(newName, ")");
                delete[] name;
              }
              name = newName;
            }
            for (vector<char*>::iterator it = extendedNames.begin(); it != extendedNames.end(); it++)
            {
              delete[] *it;
            }
          }
        }
        else if (tag == 0xa1) // service location descriptor
        {
          DecodeServiceLocationDescriptor(&section[pointer], length, &videoStreamCount, &audioStreamCount, &languages);
        }
        pointer += length;
      }

      if (m_pCallBack != NULL)
      {
        m_pCallBack->OnLvctReceived(sections.table_id, name, majorChannel, minorChannel, modulationMode, carrierFrequency,
          channelTsid, programNumber, etmLocation, accessControlled, hidden, pathSelect, outOfBand, hideGuide, serviceType,
          sourceId, videoStreamCount, audioStreamCount, languages);
      }
      if (name != NULL)
      {
        delete[] name;
      }
    }

    // additional descriptors...
    int additionalDescriptorsLength = ((section[pointer] & 0x3) << 8) + section[pointer + 1];
    pointer += 2;
    int endOfDescriptors = pointer + additionalDescriptorsLength;
    LogDebug("LvctParser: pointer = %d, additional descriptors length = %d, end of descriptors = %d", pointer, additionalDescriptorsLength, endOfDescriptors);
    while (pointer + 1 < endOfDescriptors)
    {
      int tag = section[pointer++];
      int length = section[pointer++];
      LogDebug("LvctParser: descriptor, tag = 0x%x, length = %d, pointer = %d, end of descriptor = %d", tag, length, pointer, pointer + length);
      if (pointer + length > endOfDescriptors)
      {
        LogDebug("LvctParser: invalid descriptor length = %d, pointer = %d, end of descriptor = %d, end of descriptors = %d, end of section = %d, section length = %d", length, pointer, pointer + length, endOfDescriptors, endOfSection, sectionLength);
        return;
      }

      pointer += length;
    }

    if (pointer != endOfSection)
    {
      LogDebug("LvctParser: section parsing error");
      return;
    }

    m_mUnseenSections.erase(sectionIt);
    if (m_mUnseenSections.size() == 0)
    {
      m_bIsReady = true;
      m_pCallBack = NULL;
    }
  }
  catch (...)
  {
    LogDebug("LvctParser: unhandled exception in OnNewSection()");
  }
}

void CLvctParser::DecodeServiceLocationDescriptor(byte* b, int length, int* videoStreamCount, int* audioStreamCount, vector<unsigned int>* languages)
{
  if (length < 3)
  {
    LogDebug("LvctParser: invalid service location descriptor length = %d", length);
    *videoStreamCount = 0;
    *audioStreamCount = 0;
    return;
  }
  try
  {
    int pcrPid = ((b[0] & 0x1f) << 8) + b[1];
    int numberElements = b[2];
    //LogDebug("LvctParser: PCR PID = 0x%x, number of elements = %d", pcrPid, numberElements);
    int pointer = 3;
    for (int i = 0; i < numberElements && pointer + 5 < length; i++)
    {
      int streamType = b[pointer++];
      int elementaryPid = ((b[pointer] & 0x1f) << 8) + b[pointer + 1];
      pointer += 2;
      unsigned int iso639LanguageCode = b[pointer] + (b[pointer + 1] << 8) + (b[pointer + 2] << 16);
      if (iso639LanguageCode != 0)
      {
        languages->push_back(iso639LanguageCode);
      }
      pointer += 3;
      //LogDebug("LvctParser: stream type = 0x%x, elementary PID = 0x%x", streamType, elementaryPid);

      if (streamType == SERVICE_TYPE_VIDEO_MPEG1 ||
          streamType == SERVICE_TYPE_VIDEO_MPEG2 ||
          streamType == SERVICE_TYPE_VIDEO_MPEG4 ||
          streamType == SERVICE_TYPE_VIDEO_H264 ||
          streamType == SERVICE_TYPE_VIDEO_HEVC ||
          streamType == SERVICE_TYPE_VIDEO_MPEG2_DCII)
      {
        (*videoStreamCount)++;
      }
      else if (streamType == SERVICE_TYPE_AUDIO_MPEG1 ||
          streamType == SERVICE_TYPE_AUDIO_MPEG2 ||
          streamType == SERVICE_TYPE_AUDIO_AAC ||
          streamType == SERVICE_TYPE_AUDIO_LATM_AAC ||
          streamType == SERVICE_TYPE_AUDIO_AC3 ||
          streamType == SERVICE_TYPE_AUDIO_E_AC3)
      {
        (*audioStreamCount)++;
      }
    }
    //LogDebug("LvctParser: video stream count = %d, audio stream count = %d", *videoStreamCount, *audioStreamCount);
  }
  catch (...)
  {
    LogDebug("LvctParser: unhandled exception in DecodeServiceLocationDescriptor()");
    *videoStreamCount = 0;
    *audioStreamCount = 0;
    languages->clear();
  }
}

void CLvctParser::DecodeMultipleStrings(byte* b, int length, vector<char*>* strings)
{
  if (length < 1)
  {
    LogDebug("LvctParser: invalid multiple strings structure length = %d", length);
    return;
  }
  try
  {
    int numberStrings = b[0];

    //LogDebug("LvctParser: parse multiple strings, number of strings = %d", numberStrings);
    int pointer = 1;
    for (int i = 0; i < numberStrings && pointer + 3 < length; i++)
    {
      unsigned int iso639LanguageCode = b[pointer] + (b[pointer + 1] << 8) + (b[pointer + 2] << 16);
      pointer += 3;
      int numberSegments = b[pointer++];
      //LogDebug("LvctParser: string %d, number of segments = %d", i, numberSegments);
      for (int j = 0; j < numberSegments && pointer + 2 < length; j++)
      {
        int compressionType = b[pointer++];
        int mode = b[pointer++];
        int numberBytes = b[pointer++];
        //LogDebug("LvctParser: segment %d, compression type = 0x%x, mode = 0x%x, number of bytes = %d", j, compressionType, mode, numberBytes);
        if (pointer + numberBytes > length)
        {
          LogDebug("LvctParser: invalid string length %d in multiple string structure, pointer = %d, number of bytes = %d, structure length = %d", pointer, numberBytes, length);
          return;
        }

        char* string = NULL;
        DecodeString(&b[pointer], compressionType, mode, numberBytes, &string);
        if (string != NULL)
        {
          strings->push_back(string);
        }

        pointer += numberBytes;
      }
    }
  }
  catch (...)
  {
    LogDebug("LvctParser: unhandled exception in ParseMultipleStrings()");
  }
}

void CLvctParser::DecodeString(byte* b, int compressionType, int mode, int numberBytes, char** string)
{
  //LogDebug("LvctParser: decode string, compression type = 0x%x, mode = 0x%x, number of bytes = %d", compressionType, mode, numberBytes);
  if (compressionType == 0 && mode == 0)
  {
    *string = new char[numberBytes + 1];
    if (*string == NULL)
    {
      LogDebug("LvctParser: failed to allocate %d bytes in DecodeString()", numberBytes + 1);
      return;
    }
    memcpy(*string, b, numberBytes);
    (*string)[numberBytes] = 0; // NULL terminate
    return;
  }
  LogDebug("LvctParser: unsupported compression type or mode in DecodeString(), compression type = 0x%x, mode = 0x%x", compressionType, mode);
  for (int i = 0; i < numberBytes; i++)
  {
    LogDebug(" %d: 0x%x", i, b[i]);
  }
}

File: TvEngine3/TVLibrary/TVLibrary/Implementations/Dri/Parser/LvctParser.cs
#region Copyright (C) 2005-2011 Team MediaPortal

// Copyright (C) 2005-2011 Team MediaPortal
// http://www.team-mediaportal.com
// 
// MediaPortal is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
// 
// MediaPortal is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MediaPortal. If not, see <http://www.gnu.org/licenses/>.

#endregion

using System.Collections.Generic;

namespace TvLibrary.Implementations.Dri.Parser
{
  public enum ModulationMode : byte
  {
    Analog = 0x01,
    ScteMode1 = 0x02, // 64 QAM
    ScteMode2 = 0x03, // 256 QAM
    Atsc8Vsb = 0x04,
    Atsc16Vsb = 0x05,
    PrivateDescriptor = 0x80
  }

  public enum EtmLocation : byte
  {
    None,
    PhysicalChannelThis,
    PhysicalChannelTsid
  }

  public enum AtscServiceType : byte
  {
    /// <summary>
    /// analog television (A/65)
    /// </summary>
    AnalogTelevision = 0x01,
    /// <summary>
    /// ATSC digital television service (A/53 part 3)
    /// </summary>
    DigitalTelevision = 0x02,
    /// <summary>
    /// ATSC audio service (A/53 part 3)
    /// </summary>
    Audio = 0x03,
    /// <summary>
    /// ATSC data only service (A/90)
    /// </summary>
    DataOnly = 0x04,
    /// <summary>
    /// ATSC software download service (A/97)
    /// </summary>
    SoftwareDownload = 0x05,
    /// <summary>
    /// unassociated/small screen service (A/53 part 3)
    /// </summary>
    SmallScreen = 0x06,
    /// <summary>
    /// parameterised service (A/71)
    /// </summary>
    Parameterised = 0x07,
    /// <summary>
    /// Non Real Time service (A/103)
    /// </summary>
    Nrt = 0x08,
    /// <summary>
    /// extended parametised service (A/71)
    /// </summary>
    ExtendedParameterised = 0x09
  }

  public delegate void LvctChannelDetailDelegate(MgtTableType tableType, string shortName, int majorChannelNumber, int minorChannelNumber,
      ModulationMode modulationMode, uint carrierFrequency, int channelTsid, int programNumber, EtmLocation etmLocation,
      bool accessControlled, bool hidden, int pathSelect, bool outOfBand, bool hideGuide, AtscServiceType serviceType, int sourceId);

  /// <summary>
  /// ATSC/SCTE long form virtual channel table parser. Refer to ATSC A-65 and SCTE 65.
  /// </summary>
  public class LvctParser
  {
    private int _currentVersion = -1;
    private HashSet<int> _unseenSections = new HashSet<int>();
    public event TableCompleteDelegate OnTableComplete = null;
    public event LvctChannelDetailDelegate OnChannelDetail = null;

    public void Reset()
    {
      _currentVersion = -1;
    }

    public void Decode(byte[] section)
    {
      if (OnTableComplete == null)
      {
        return;
      }
      if (section.Length < 18)
      {
        Log.Log.Error("L-VCT: invalid section size {0}, expected at least 18 bytes", section.Length);
        return;
      }

      byte tableId = section[2];
      // 0xc8 = ATSC terrestrial, A-65
      // 0xc9 = SCTE cable, SCTE-65
      // The cable and terrestrial L-VCT formats are almost identical. The few
      // differences are noted below.
      if (tableId != 0xc8 && tableId != 0xc9)
      {
        return;
      }
      MgtTableType tableType = MgtTableType.CvctCurrentNext1;
      if (tableId == 0xc8)
      {
        tableType = MgtTableType.TvctCurrentNext1;
      }
      bool sectionSyntaxIndicator = ((section[3] & 0x80) != 0);
      bool privateIndicator = ((section[3] & 0x40) != 0);
      int sectionLength = ((section[3] & 0x0f) << 8) + section[4];
      if (section.Length != 2 + sectionLength + 3)  // 2 for section length bytes, 3 for table ID and PID
      {
        Log.Log.Error("L-VCT: invalid section length = {0}, byte count = {1}", sectionLength, section.Length);
        return;
      }
      int transportStreamId = (section[5] << 8) + section[6];
      int versionNumber = ((section[7] >> 1) & 0x1f);
      bool currentNextIndicator = ((section[7] & 0x80) != 0);
      if (!currentNextIndicator)
      {
        // Not applicable yet.
        return;
      }
      byte sectionNumber = section[8];
      byte lastSectionNumber = section[9];
      int sectionKey = (tableId << 8) + sectionNumber;
      if (versionNumber > _currentVersion || (_currentVersion == 31 && versionNumber < _currentVersion))
      {
        _currentVersion = versionNumber;
        _unseenSections.Clear();
        for (int s = 0; s <= lastSectionNumber; s++)
        {
          _unseenSections.Add((tableId << 8) + s);
        }
      }
      else if (!_unseenSections.Contains(sectionKey))
      {
        // Already seen this section.
        return;
      }

      byte protocolVersion = section[10];
      int numChannelsInSection = section[11];
      Log.Log.Debug("L-VCT: section length = {0}, transport stream ID = 0x{1:x}, version number = {2}, section number = {3}, last section number {4}, protocol version = {5}, number of channels in section = 0x{6:x}",
        sectionLength, transportStreamId, versionNumber, sectionNumber, lastSectionNumber, protocolVersion, numChannelsInSection);

      int pointer = 12;
      int endOfSection = section.Length - 4;
      for (int i = 0; i < numChannelsInSection; i++)
      {
        if (pointer + 32 + 2 > endOfSection)  // + 2 for the fixed bytes after the loop
        {
          Log.Log.Error("L-VCT: detected number of channels in section {0} is invalid, pointer = {1}, end of section = {2}, loop = {3}", numChannelsInSection, pointer, endOfSection, i);
          return;
        }

        string shortName = System.Text.Encoding.Unicode.GetString(section, pointer, 14);
        pointer += 14;

        // SCTE cable supports both one-part and two-part channel numbers where
        // the major and minor channel number range is 0..999.
        // ATSC supports only two-part channel numbers where the major range is
        // 1..99 and the minor range is 0..999.
        // When the minor channel number is 0 it indicates an analog channel.
        int majorChannelNumber = ((section[pointer] & 0x0f) << 6) + (section[pointer + 1] >> 2);
        pointer++;
        int minorChannelNumber = ((section[pointer] & 0x03) << 8) + section[pointer + 1];
        pointer += 2;
        if ((majorChannelNumber & 0x03f0) == 0x03f0)
        {
          majorChannelNumber = ((majorChannelNumber & 0x0f) << 10) + minorChannelNumber;
          minorChannelNumber = 0;
        }

        ModulationMode modulationMode = (ModulationMode)section[pointer++];
        uint carrierFrequency = 0;   // Hz
        for (byte b = 0; b < 4; b++)
        {
          carrierFrequency = carrierFrequency << 8;
          carrierFrequency = section[pointer++];
        }
        int channelTsid = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        int programNumber = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        EtmLocation etmLocation = (EtmLocation)(section[pointer] >> 6);   // ATSC only, SCTE reserved
        bool accessControlled = ((section[pointer] & 0x20) != 0);
        bool hidden = ((section[pointer] & 0x10) != 0);
        int pathSelect = ((section[pointer] & 0x08) >> 3);                // SCTE only, ATSC reserved
        bool outOfBand = ((section[pointer] & 0x04) != 0);                // SCTE only, ATSC reserved
        bool hideGuide = ((section[pointer++] & 0x02) != 0);
        AtscServiceType serviceType = (AtscServiceType)(section[pointer++] & 0x3f);
        int sourceId = (section[pointer] << 8) + section[pointer + 1];
        pointer += 2;
        Log.Log.Debug("L-VCT: channel, short name = {0}, major channel number = {1}, minor channel number = {2}, modulation mode = {3}, carrier frequency = {4} Hz, TSID = 0x{5:x}, program number = 0x{6:x}, ETM location = {7}, access controlled = {8}, hidden = {9}, path select = {10}, out of band = {11}, hide guide = {12}, service type = {13}, source ID = 0x{14:x}",
          shortName, majorChannelNumber, minorChannelNumber, modulationMode, carrierFrequency, channelTsid, programNumber,
          etmLocation, accessControlled, hidden, pathSelect, outOfBand, hideGuide, serviceType, sourceId);

        if (OnChannelDetail != null)
        {
          OnChannelDetail(tableType, shortName, majorChannelNumber, minorChannelNumber, modulationMode, carrierFrequency, channelTsid,
            programNumber, etmLocation, accessControlled, hidden, pathSelect, outOfBand, hideGuide, serviceType, sourceId);
        }

        int descriptorsLength = ((section[pointer] & 0x03) << 8) + section[pointer + 1];
        pointer += 2;
        int endOfDescriptors = pointer + descriptorsLength;
        if (endOfDescriptors > endOfSection)
        {
          Log.Log.Error("L-VCT: invalid descriptors length {0}, pointer = {1}, end of section = {2}, loop = {3}", descriptorsLength, pointer, endOfSection, i);
          return;
        }
        while (pointer + 1 < endOfDescriptors)
        {
          byte tag = section[pointer++];
          byte length = section[pointer++];
          Log.Log.Debug("L-VCT: descriptor, tag = 0x{0:x}, length = {1}", tag, length);
          if (pointer + length > endOfDescriptors)
          {
            Log.Log.Error("L-VCT: invalid descriptor length {0}, pointer = {1}, end of descriptors = {2}, loop = {3}", length, pointer, endOfDescriptors, i);
            return;
          }
          pointer += length;
        }
        if (pointer != endOfDescriptors)
        {
          Log.Log.Error("L-VCT: corruption detected at end of descriptors, pointer = {0}, end of section = {1}, end of descriptors = {2}, loop = {3}", pointer, endOfSection, endOfDescriptors, i);
          return;
        }
      }

      int additionalDescriptorsLength = ((section[pointer] & 0x03) << 8) + section[pointer + 1];
      pointer += 2;
      if (pointer + additionalDescriptorsLength != endOfSection)
      {
        Log.Log.Error("L-VCT: invalid additional descriptors length {0}, pointer = {1}, end of section = {2}", additionalDescriptorsLength, pointer, endOfSection);
        return;
      }
      while (pointer + 1 < endOfSection)
      {
        byte tag = section[pointer++];
        byte length = section[pointer++];
        Log.Log.Debug("L-VCT: additional descriptor, tag = 0x{0:x}, length = {1}", tag, length);
        if (pointer + length > endOfSection)
        {
          Log.Log.Error("L-VCT: invalid additional descriptor length {0}, pointer = {1}, end of section = {2}", length, pointer, endOfSection);
          return;
        }
        pointer += length;
      }

      if (pointer != endOfSection)
      {
        Log.Log.Error("L-VCT: corruption detected at end of section, pointer = {0}, end of section = {1}", pointer, endOfSection);
        return;
      }

      _unseenSections.Remove(sectionKey);
      if (_unseenSections.Count == 0 && OnTableComplete != null)
      {
        OnTableComplete(tableType);
        OnTableComplete = null;
        OnChannelDetail = null;
      }
    }
  }
}
